### ISSUE TITLE
Alert Transport Test OK but real life alert not working

### ISSUE THREAD

**Issue by Djf2884** [librenms/librenms#2856]
Hello, 

I m writting here because i m starting to be out of idea.

I m trying to setup the librenms alerting system, rules has been setup and all configuration is set.

Test transport is working with mail/sendmail or smtp. 

When i shutdown one of my system and when an alter is trigger, i receive the following warning on librenms : 

Could not issue critical alert for rule 'Devices up/down' to transport 'mail'

I have actually no idea how to resolve this issue as all is working, i can send email from the system, test transport is working...

I've tried to reinstall librenms but i face the same issue.

I there anyone that could have an idea or help me?

Ps : my system is raspbian lite running on a raspberry pi2.

Thx u.

**Comment by laf** [comment]
Can you screenshot the settings from the webui? I'm expecting maybe you've not got the options set to who to send mails to.

**Comment by Djf2884** [comment]
Yes sure :) 

Here it is : 

![image](https://cloud.githubusercontent.com/assets/10075667/12554595/93013d92-c37d-11e5-8295-3abbadf887bf.png)

![image](https://cloud.githubusercontent.com/assets/10075667/12554627/b6359f88-c37d-11e5-84f3-74c44f607b8c.png)

**Comment by laf** [comment]
I'm using exactly the same setup and it works fine. Can you confirm the email address for Default contact is correct?

You say if you click test transport on mail that works ok? That's basically the same code.

Can you screenshot your alert rules please?

**Comment by ekoyle** [comment]
I've noticed that our install recently stopped sending email, although
slack alerts are working (the email address lines in the slack alerts are
empty).

**Comment by Djf2884** [comment]
Hello :) Sorry for the delay, it was night here in France, below is a screenshot of my rules (default one for now)

![image](https://cloud.githubusercontent.com/assets/10075667/12573775/6566811e-c3fd-11e5-9d08-53222caddc9e.png)

Also for ur information, i have done some more testing with different tools : ssmtp, sendmail and postfix, none of them work but test transport is working in every case with every type (mail smtp and sendmail).

During test transport i can see that mail.log file is properly fill and all the logs are ok but during real life testing, i can't see any log in the mail.log file so i guess that it's not arriving to the relay.

**Comment by f0o** [comment]
Sounds like a perms issue to me. The test function us being executed as httpd or www-data depending on ur distro and if using selinux from an httpd context whereas the actual alert is being issued by the librenms user (or root, depending on your setup) and from the cron context (selinux).

Verify that selinux is permissive and that the librenms user (if any) can use the utility you're attempting to use (exec sendmail, open sockets, ..)

On 26 January 2016 07:23:09 CET, Djf2884 <notifications@github.com> wrote:
>Hello :) Sorry for the delay, it was night here in France, below is a
>screenshot of my rules (default one for now)
>
>![image](https://cloud.githubusercontent.com/assets/10075667/12573775/6566811e-c3fd-11e5-9d08-53222caddc9e.png)
>
>Also for ur information, i have done some more testing with different
>tools : ssmtp, sendmail and postfix, none of them work but test
>transport is working in every case with every type (mail smtp and
>sendmail).
>
>During test transport i can see that mail.log file is properly fill and
>all the logs are ok but during real life testing, i can't see any log
>in the mail.log file so i guess that it's not arriving to the relay.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/librenms/librenms/issues/2856#issuecomment-174852185

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Djf2884** [comment]
Hello,

By default selinux is not enable in debian/raspbian.

Also, as said, I have done multiple testing with postfix, sendmail and ssmtp. Every time, I'm able to send email with mail -s command with every users on my system including librenms and www-data one.

Test transport also working fine. 

Is there any way to get some debug log from librenms itself that could help ?

**Comment by f0o** [comment]
Sure add `2>&1 >>/path/to/alert.log` after the alerts.php cronjob

**Comment by Djf2884** [comment]
Thx I will do some more testing this morning and will get back to u with the result.

**Comment by Djf2884** [comment]
Here are the logs :

Start: Tue, 26 Jan 2016 08:51:01 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:51:01 +0100
Start: Tue, 26 Jan 2016 08:52:02 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:52:02 +0100
Start: Tue, 26 Jan 2016 08:53:01 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:53:01 +0100
Start: Tue, 26 Jan 2016 08:54:01 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:54:02 +0100
Start: Tue, 26 Jan 2016 08:55:01 +0100
RunFollowUp():
RunAlerts():
Issuing Alert-UID #114/1: ; ; mail => ERROR; ; ; ; ; ; ;
RunAcks():
End  : Tue, 26 Jan 2016 08:55:01 +0100
Start: Tue, 26 Jan 2016 08:56:02 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:56:02 +0100
Start: Tue, 26 Jan 2016 08:57:01 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:57:01 +0100
Start: Tue, 26 Jan 2016 08:58:02 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:58:02 +0100
Start: Tue, 26 Jan 2016 08:59:01 +0100
RunFollowUp():
RunAlerts():
RunAcks():
End  : Tue, 26 Jan 2016 08:59:01 +0100

Nothing clear to me but we can see that there a mail error.

**Comment by laf** [comment]
I've just updated to e65fd24fff30606b650fe8394bb88b62062f0053 version which was from yesterday and this works perfectly fine for email alerts still for me :/

Can you provide the output of ./validate.php

@f0o If we merge the debug PR in then we can get more info from alerts.php on this one.

**Comment by laf** [comment]
lol. That's been merged.

@Djf2884 if you can run ./daily.sh then provide the output of ./validate.php

Also, update the cron job to do logging as before but at -d after alerts.php (alerts.php -d) so we can get more data back.

**Comment by Djf2884** [comment]
Here is the output of daily.sh :

/opt/librenms/daily.sh
Updating to latest codebase                        OK
Updating SQL-Schema                                OK
Updating submodules                                OK
Cleaning up DB                                     OK
Fetching notifications                             OK

Output of validate.php : 
clockmodders@clmlrpi:~ $ sudo /opt/librenms/validate.php
[FAIL]    Syntax error in config.php
fatal: Not a git repository (or any of the parent directories): .git
Version info:
[WARN]    Your install is out of date:
DB Schema: 94
PHP: 5.6.17-0+deb8u1
MySQL: 10.0.22-MariaDB-0+deb8u1
RRDTool: 1.4.8
SNMP: NET-SNMP version: 5.7.2.1

[FAIL]    We have found some files that are owned by a different user than librenms, this will stop you updating automatically and / or rrd files being updated causing graphs to fail:

/opt/librenms/logs
/opt/librenms/logs/access_log
/opt/librenms/logs/error_log

[OK]      Database connection successful
[OK]      fping has been found

output result of validate.php seem strange to me as i have reinstall all frm scratch yesterday..

here are the debug logs :

DEBUG!
Start: Tue, 26 Jan 2016 09:36:01 +0100
RunFollowUp():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.state > 0 && alerts.open = 0] RunAlerts():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.open = 1]
SQL[SELECT alert_log.id,alert_log.rule_id,alert_log.device_id,alert_log.state,alert_log.details,alert_log.time_logged,alert_rules.rule,alert_rules.severity,alert_rules.extra,alert_rules.name FROM alert_log,alert_rules WHERE alert_log.rule_id = alert_rules.id && alert_log.device_id = '8' && alert_log.rule_id = '1' && alert_rules.disabled = 0 ORDER BY alert_log.id DESC LIMIT 1]
SQL[SELECT * FROM device_groups ORDER BY name]
SQL[SELECT alert_rules.* FROM alert_rules LEFT JOIN alert_map ON alert_rules.id=alert_map.rule WHERE alert_rules.disabled = 0 && ( (alert_rules.device_id = -1 || alert_rules.device_id = '8' ) || alert_map.target = '8'  )]
SQL[SELECT alerts.alerted,devices.ignore,devices.disabled FROM alerts,devices WHERE alerts.device_id = '8' && devices.device_id = alerts.device_id && alerts.rule_id = '1']
SQL[SELECT * FROM device_groups ORDER BY name]
SQL[SELECT alert_schedule.schedule_id FROM alert_schedule LEFT JOIN alert_schedule_items ON alert_schedule.schedule_id=alert_schedule_items.schedule_id WHERE ( alert_schedule_items.target = '8' ) && NOW() BETWEEN alert_schedule.start AND alert_schedule.end LIMIT 1]
SQL[UPDATE `alert_log` set `details` ='xګVJI▒I▒T▒24156▒4763▒\0?A▒' WHERE id = '114'] xtermxterm
SQL[SELECT attrib_value FROM devices_attribs WHERE attrib_type = "disable_notify" && device_id = '8']
SQL[SELECT hostname FROM devices WHERE device_id = '8']
SQL[SELECT `template`,`title`,`title_rec` FROM `alert_templates` JOIN `alert_template_map` ON `alert_template_map`.`alert_templates_id`=`alert_templates`.`id` WHERE `alert_template_map`.`alert_rule_id`='1'] Issuing Alert-UID #114/1: ; ; mail => ERROR
SQL[SELECT * FROM `devices` WHERE `device_id` = '8']
SQL[INSERT INTO `eventlog` (`host`,`reference`,`type`,`datetime`,`message`)  VALUES ('8','NULL','NULL',NOW(),'Could not issue critical alert for rule \'Devices up/down\' to transport \'mail\'')] ; ; ; ; ; ; ;

SQL[UPDATE `alerts` set `alerted` ='1' WHERE rule_id = '1' && device_id = '8'] RunAcks():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state = 2 && alerts.open = 1] End  : Tue, 26 Jan 2016 09:36:04 +0100
DEBUG!
Start: Tue, 26 Jan 2016 09:37:02 +0100
RunFollowUp():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.state > 0 && alerts.open = 0] RunAlerts():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.open = 1]
SQL[SELECT alert_log.id,alert_log.rule_id,alert_log.device_id,alert_log.state,alert_log.details,alert_log.time_logged,alert_rules.rule,alert_rules.severity,alert_rules.extra,alert_rules.name FROM alert_log,alert_rules WHERE alert_log.rule_id = alert_rules.id && alert_log.device_id = '8' && alert_log.rule_id = '1' && alert_rules.disabled = 0 ORDER BY alert_log.id DESC LIMIT 1]
SQL[SELECT * FROM device_groups ORDER BY name]
SQL[SELECT alert_rules.* FROM alert_rules LEFT JOIN alert_map ON alert_rules.id=alert_map.rule WHERE alert_rules.disabled = 0 && ( (alert_rules.device_id = -1 || alert_rules.device_id = '8' ) || alert_map.target = '8'  )]
SQL[SELECT alerts.alerted,devices.ignore,devices.disabled FROM alerts,devices WHERE alerts.device_id = '8' && devices.device_id = alerts.device_id && alerts.rule_id = '1'] RunAcks():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state = 2 && alerts.open = 1] End  : Tue, 26 Jan 2016 09:37:02 +0100

**Comment by laf** [comment]
Please run ./validate.php from within the /opt/librenms directory - if you did then the config.php issue could explain it but I'd expect it all to break at that stage.

**Comment by Djf2884** [comment]
here is the new run regarding validate.php :

clockmodders@clmlrpi:/opt/librenms $ sudo ./validate.php
Version info:
Commit SHA: 3e97875d77fd02104dd6060f6260e394adaea673
DB Schema: 94
PHP: 5.6.17-0+deb8u1
MySQL: 10.0.22-MariaDB-0+deb8u1
RRDTool: 1.4.8
SNMP: NET-SNMP version: 5.7.2.1

[FAIL]    We have found some files that are owned by a different user than librenms, this will stop you updating automatically and / or rrd files being updated causing graphs to fail:

/opt/librenms/logs
/opt/librenms/logs/access_log
/opt/librenms/logs/error_log

[OK]      Database connection successful
[OK]      fping has been found

**Comment by Djf2884** [comment]
I just notice that the config.php file start with <?php but there is no ?> to close the quote at the end.

Is that normal? in default config file i have the exact same case.

Anyway after adding the ?> at the end, the issue is still the same.. Any other idea would be welcome :D

**Comment by laf** [comment]
That's normal. Please don't add ?> to the end.

"Also, update the cron job to do logging as before but at -d after alerts.php (alerts.php -d) so we can get more data back."

**Comment by Djf2884** [comment]
already sent before, do u want me to send u more logs than the one previously sent : 

DEBUG!
Start: Tue, 26 Jan 2016 09:36:01 +0100
RunFollowUp():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.state > 0 && alerts.open = 0] RunAlerts():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.open = 1]
SQL[SELECT alert_log.id,alert_log.rule_id,alert_log.device_id,alert_log.state,alert_log.details,alert_log.time_logged,alert_rules.rule,alert_rules.severity,alert_rules.extra,alert_rules.name FROM alert_log,alert_rules WHERE alert_log.rule_id = alert_rules.id && alert_log.device_id = '8' && alert_log.rule_id = '1' && alert_rules.disabled = 0 ORDER BY alert_log.id DESC LIMIT 1]
SQL[SELECT * FROM device_groups ORDER BY name]
SQL[SELECT alert_rules.* FROM alert_rules LEFT JOIN alert_map ON alert_rules.id=alert_map.rule WHERE alert_rules.disabled = 0 && ( (alert_rules.device_id = -1 || alert_rules.device_id = '8' ) || alert_map.target = '8' )]
SQL[SELECT alerts.alerted,devices.ignore,devices.disabled FROM alerts,devices WHERE alerts.device_id = '8' && devices.device_id = alerts.device_id && alerts.rule_id = '1']
SQL[SELECT * FROM device_groups ORDER BY name]
SQL[SELECT alert_schedule.schedule_id FROM alert_schedule LEFT JOIN alert_schedule_items ON alert_schedule.schedule_id=alert_schedule_items.schedule_id WHERE ( alert_schedule_items.target = '8' ) && NOW() BETWEEN alert_schedule.start AND alert_schedule.end LIMIT 1]
SQL[UPDATE alert_log set details ='xګVJI▒I▒T▒24156▒4763▒\0?A▒' WHERE id = '114'] xtermxterm
SQL[SELECT attrib_value FROM devices_attribs WHERE attrib_type = "disable_notify" && device_id = '8']
SQL[SELECT hostname FROM devices WHERE device_id = '8']
SQL[SELECT template,title,title_rec FROM alert_templates JOIN alert_template_map ON alert_template_map.alert_templates_id=alert_templates.id WHERE alert_template_map.alert_rule_id='1'] Issuing Alert-UID #114/1: ; ; mail => ERROR
SQL[SELECT * FROM devices WHERE device_id = '8']
SQL[INSERT INTO eventlog (host,reference,type,datetime,message) VALUES ('8','NULL','NULL',NOW(),'Could not issue critical alert for rule \'Devices up/down\' to transport \'mail\'')] ; ; ; ; ; ; ;

SQL[UPDATE alerts set alerted ='1' WHERE rule_id = '1' && device_id = '8'] RunAcks():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state = 2 && alerts.open = 1] End : Tue, 26 Jan 2016 09:36:04 +0100
DEBUG!
Start: Tue, 26 Jan 2016 09:37:02 +0100
RunFollowUp():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.state > 0 && alerts.open = 0] RunAlerts():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state != 2 && alerts.open = 1]
SQL[SELECT alert_log.id,alert_log.rule_id,alert_log.device_id,alert_log.state,alert_log.details,alert_log.time_logged,alert_rules.rule,alert_rules.severity,alert_rules.extra,alert_rules.name FROM alert_log,alert_rules WHERE alert_log.rule_id = alert_rules.id && alert_log.device_id = '8' && alert_log.rule_id = '1' && alert_rules.disabled = 0 ORDER BY alert_log.id DESC LIMIT 1]
SQL[SELECT * FROM device_groups ORDER BY name]
SQL[SELECT alert_rules.* FROM alert_rules LEFT JOIN alert_map ON alert_rules.id=alert_map.rule WHERE alert_rules.disabled = 0 && ( (alert_rules.device_id = -1 || alert_rules.device_id = '8' ) || alert_map.target = '8' )]
SQL[SELECT alerts.alerted,devices.ignore,devices.disabled FROM alerts,devices WHERE alerts.device_id = '8' && devices.device_id = alerts.device_id && alerts.rule_id = '1'] RunAcks():

SQL[SELECT alerts.device_id, alerts.rule_id, alerts.state FROM alerts WHERE alerts.state = 2 && alerts.open = 1] End : Tue, 26 Jan 2016 09:37:02 +0100

**Comment by f0o** [comment]
Ok just to revive an old possibility:
Have you changed the email recepient (either by enabling or modifying the default_contact or the syscontact or even your accounts email) *after* the alert got triggered?
These changes arent processed for ongoing alerts and only affect new alerts.
There is a config option to re-evaluate contact changes before sending alerts, cant remember the name but it's in the UI under general settings.

**Comment by Djf2884** [comment]
Hello Guys :)

I finaly finaly found the trick !

Yesterday after the third librenms reinstallation, i decide to reinstall the full system !, so i follow the librenms documentation again and working after configuration without any problem !

For documentation purpose i m recording every command on my systems during deployment so i decide to review the first installation to see the difference and i put the finger on it !

During the first system install i forgot to uncomment date.timezone = Europe/Paris in php.ini file.

I have proceed to the librenms installation and after the deployment i have uncoment the date.timezone in the php.ini.

I don't know if it's possible but i think this should be verified before librenms deployment through the GUi. Also i don't know why uncomenting the this after librenms install is causing so much trouble.

I have restored the original system uncoment the line and redeploy librenms and it was working like a charm !

Thx u all for ur help, i really appreciate it :)
